exp/operator/controllers/shared_test.go

```diff
@@ -0,0 +1,171 @@
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "core-1",
+						Namespace: namespaceName1,
+					},
+					TypeMeta: metav1.TypeMeta{
+						Kind:       "CoreProvider",
+						APIVersion: "operator.cluster.x-k8s.io/v1alpha4",
+					},
+				},
+				&operatorv1.CoreProvider{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "core-2",
+						Namespace: namespaceName1,
+					},
+					TypeMeta: metav1.TypeMeta{
+						Kind:       "CoreProvider",
+						APIVersion: "operator.cluster.x-k8s.io/v1alpha4",
+					},
+				},
+			},
+			expectedCondition: clusterv1.Condition{
+				Type:     operatorv1.PreflightCheckCondition,
+				Reason:   operatorv1.MoreThanOneExistsReason,
+				Severity: clusterv1.ConditionSeverityWarning,
+				Message:  operatorv1.MoreThanOneExistsMessage,
+				Status:   corev1.ConditionFalse,
+			},
+		},
+		{
+			name: "two core providers in two different namespaces were created, preflight check failed",
+			providers: []client.Object{
+				&operatorv1.CoreProvider{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "core-1",
+						Namespace: namespaceName1,
+					},
+					TypeMeta: metav1.TypeMeta{
+						Kind:       "CoreProvider",
+						APIVersion: "operator.cluster.x-k8s.io/v1alpha4",
+					},
+				},
+				&operatorv1.CoreProvider{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "core-2",
+						Namespace: namespaceName2,
+					},
+					TypeMeta: metav1.TypeMeta{
+						Kind:       "CoreProvider",
+						APIVersion: "operator.cluster.x-k8s.io/v1alpha4",
+					},
+				},
+			},
+			expectedCondition: clusterv1.Condition{
+				Type:     operatorv1.PreflightCheckCondition,
+				Reason:   operatorv1.MoreThanOneExistsReason,
+				Severity: clusterv1.ConditionSeverityWarning,
+				Message:  operatorv1.MoreThanOneExistsMessage,
+				Status:   corev1.ConditionFalse,
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			gs := NewWithT(t)
+
+			gvk := tc.providers[0].GetObjectKind().GroupVersionKind()
+
+			for _, c := range tc.providers {
+				gs.Expect(testEnv.Create(ctx, c)).To(Succeed())
+			}
+
+			actualCondition, err := preflightChecks(context.Background(), testEnv.Client, gvk)
+			gs.Expect(err).ToNot(HaveOccurred())
+
+			// Check if proper condition is returned
+			gs.Expect(actualCondition.Message).To(Equal(tc.expectedCondition.Message))
+			gs.Expect(actualCondition.Type).To(Equal(tc.expectedCondition.Type))
+			gs.Expect(actualCondition.Status).To(Equal(tc.expectedCondition.Status))
+			gs.Expect(actualCondition.Severity).To(Equal(tc.expectedCondition.Severity))
+
+			for _, c := range tc.providers {
+				gs.Expect(testEnv.Delete(ctx, c)).To(Succeed())
+				// Wait for CoreProvider to be deleted.
+				g.Eventually(func() bool {
+					if err := testEnv.Get(ctx, client.ObjectKeyFromObject(c), c); err != nil {
+						if apierrors.IsNotFound(err) {
+							return true
+						}
+					}
+					return false
+				}, timeout).Should(BeTrue())
+			}
```

Comment:
Can we use the `testEnv.Cleanup` to get rid of these? I'm concerned that having these after the expectations above will mean that they won't get cleaned up in case of failure.

Would be better to defer cleanup or deletion at the point each resource is created I think

Reply:
testEnv.Cleanup() doesn't seem to wait for object to be deleted